l1-contracts/contracts/governance/ServerNotifier.sol

```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.0;
+
+import {Ownable2Step} from "@openzeppelin/contracts-v4/access/Ownable2Step.sol";
+import {Initializable} from "@openzeppelin/contracts-v4/proxy/utils/Initializable.sol";
+import {ZeroAddress, Unauthorized} from "../common/L1ContractErrors.sol";
+import {ReentrancyGuard} from "../common/ReentrancyGuard.sol";
+import {IChainTypeManager} from "../state-transition/IChainTypeManager.sol";
+
+contract ServerNotifier is Ownable2Step, ReentrancyGuard, Initializable {
+    IChainTypeManager public chainTypeManager;
+
+    event MigrateToGateway(uint256 indexed chainId);
+    event MigrateFromGateway(uint256 indexed chainId);
+
+    /// @notice Checks if the caller is the admin of the chain.
+    modifier onlyChainAdmin(uint256 _chainId) {
+        if (msg.sender != chainTypeManager.getChainAdmin(_chainId)) {
+            revert Unauthorized(msg.sender);
+        }
+        _;
+    }
+
+    constructor(bool disableInitializers) {
+        if (disableInitializers) {
+            _disableInitializers();
+        }
+    }
+
+    function initialize(address _admin) public reentrancyGuardInitializer {
+        if (_admin == address(0)) {
+            revert ZeroAddress();
+        }
+
+        _transferOwnership(_admin);
+    }
+
+    function setChainTypeManager(IChainTypeManager _chainTypeManager) external onlyOwner {
+        if (address(_chainTypeManager) == address(0)) {
+            revert ZeroAddress();
+        }
+        chainTypeManager = IChainTypeManager(_chainTypeManager);
+    }
+
+    function migrateToGateway(uint256 _chainId) external onlyChainAdmin(_chainId) {
```

Comment:
What happens if chain admin calls this just for fun? Maybe even several times?

Reply:
And what happens if migration started _without_ this event?

Reply:
If chain admins run it for fun. it will stop the commitBlock production. 
If started without an event it will lead to probability of non correct root hash on gateway

Reply:
So basically it creates some attack vectors

Reply:
I think it should check, that migration isn't paused in Bridgehub https://github.com/matter-labs/era-contracts/blob/e347bf0f57fbc23ceaf9d348c61398de97e494d6/l1-contracts/contracts/bridgehub/IBridgehub.sol#L117

Reply:
And ideally migration should not be possible without this event. So imo, it actually makes sense to move all this functionality inside BridgeHub itself.